File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/clustering/partitional/ByCentroids.java
package cujae.inf.ic.om.heuristic.assignment.clustering.partitional;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;

import cujae.inf.ic.om.factory.DistanceType;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;
import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Depot;
import cujae.inf.ic.om.problem.input.Fleet;
import cujae.inf.ic.om.problem.input.Location;
import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.output.solution.Cluster;

public abstract class ByCentroids extends Partitional {
	
	protected ArrayList<Integer> generateElements(DistanceType distanceType){
		
		ArrayList<Integer> idElements = new ArrayList<Integer>();
		
		int totalCustomers = Problem.getProblem().getTotalCustomers();
		int totalDepots = Problem.getProblem().getTotalDepots();
		int counter = totalDepots;

		NumericMatrix costMatrix = null;
		RowCol rcBestAll = new RowCol();
		int idElement = -1 ;
			
		Depot depot = new Depot();
		depot.setIDDepot(-1);
		depot.setLocationDepot(calculateMeanCoordinate());
		
		ArrayList<Depot> listDepot = new ArrayList<Depot>();
		listDepot.add(depot);
		
		switch(distanceType.ordinal())
		{
			case 0: case 1: case 2: case 3:
			{
				try {
					costMatrix = Problem.getProblem().fillCostMatrix(Problem.getProblem().getCustomers(), listDepot, distanceType);
				} catch (IllegalArgumentException | SecurityException | ClassNotFoundException | InstantiationException
						| IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
					e.printStackTrace();
				}
	
				break;
			}
			case 4:
			{
				//costMatrix = new NumericMatrix(InfoProblem.getProblem().getCostMatrix()); // NO FUNcIONA
	
				break;
			}
		}
		
		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + idElements);

		while(counter > 0)
		{
			rcBestAll = costMatrix.indexBiggerValue(totalCustomers, 0, totalCustomers, (totalCustomers - 1));

			System.out.println("ROW SELECCIONADA: " + rcBestAll.getRow());
			System.out.println("COL SELECCIONADA: " + rcBestAll.getCol());
			System.out.println("VALOR SELECCIONADO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));

			idElement = Problem.getProblem().getCustomers().get(rcBestAll.getCol()).getIDCustomer();	
			idElements.add(idElement);	

			System.out.println("ELEMENTO: " + idElement); 
			System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);

			costMatrix.setItem(rcBestAll.getRow(), rcBestAll.getCol(), Double.NEGATIVE_INFINITY);
			counter--;
		}
		
		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + idElements);
	
		return sortedElements(idElements, distanceType);
	}
	
	private ArrayList<Integer> sortedElements (ArrayList<Integer> idElements, DistanceType distanceType){
		
		int totalDepots = Problem.getProblem().getTotalDepots();
		int j = 0;
		
		ArrayList<Integer> sortedElements = new ArrayList<Integer>();
		ArrayList<Customer> customers = new ArrayList<Customer>();

		for(int i = 0; i < idElements.size(); i++)
		{
			sortedElements.add(-1);
			customers.add(Problem.getProblem().getCustomerByIDCustomer(idElements.get(i)));
		}
		
		NumericMatrix costMatrix = null;
		RowCol rcBestAll = new RowCol();
		
		switch(distanceType.ordinal())
		{
			case 0: case 1: case 2: case 3:
			{
				try {
					costMatrix = Problem.getProblem().fillCostMatrix(customers, Problem.getProblem().getDepots(), distanceType);
				} catch (IllegalArgumentException | SecurityException | ClassNotFoundException | InstantiationException
						| IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
					e.printStackTrace();
				}
	
				break;
			}
			case 4:
			{
				//costMatrix = new NumericMatrix(InfoProblem.getProblem().getCostMatrix());

				break;
			}
		}
		
//		System.out.println("----ORGANIZAR ELEMENTOS SELECCIONADOS-----------------------------------------------------");
		
		while(j < idElements.size())
		{
			rcBestAll = costMatrix.indexLowerValue(0, idElements.size(), (idElements.size() - 1), (idElements.size() + totalDepots - 1));
			
			System.out.println("ROW SELECCIONADA: " + rcBestAll.getRow());
			System.out.println("COL SELECCIONADA: " + rcBestAll.getCol());
			System.out.println("VALOR SELECCIONADO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));
			
			costMatrix.fillValue(0, rcBestAll.getCol(), (idElements.size() - 1), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
			costMatrix.fillValue(rcBestAll.getRow(), (idElements.size() - 1), rcBestAll.getRow(), (idElements.size() + totalDepots - 1), Double.POSITIVE_INFINITY);
			
			sortedElements.set((rcBestAll.getCol() - idElements.size()), idElements.get(rcBestAll.getRow()));
			
			System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS ORDENADOS ACTUALIZADA:" + sortedElements);
			
			j++;
		}
		
		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS ORDENADOS:" + sortedElements);
		
		return sortedElements;
	}
	
	private Location calculateMeanCoordinate(){
		
		double AxisX = 0.0;
		double AxisY = 0.0;

		ArrayList<Location> listCoordinatesCustomers = Problem.getProblem().getListCoordinatesCustomers();

		for(int i = 0; i < listCoordinatesCustomers.size(); i++) 
		{
			AxisX += listCoordinatesCustomers.get(i).getAxisX();
			AxisY += listCoordinatesCustomers.get(i).getAxisY();
		}

		AxisX /= listCoordinatesCustomers.size();
		AxisY /= listCoordinatesCustomers.size();

		Location meanLocation = new Location(AxisX, AxisY);
		
		return meanLocation;
	}
	
	protected boolean verifyCentroids(ArrayList<Cluster> clusters, ArrayList<Depot> centroids, DistanceType distanceType) {
		boolean change = false;
		Location dummyDepot;

		System.out.println("change: " + change);
		
		for(int i = 0; i < clusters.size(); i++) 
		{
			if(!clusters.get(i).getItemsOfCluster().isEmpty())
				dummyDepot = recalculateCentroid(clusters.get(i));
			else
				dummyDepot = centroids.get(i).getLocationDepot(); 
			
			System.out.println("------------------------------------------------------------------");
			System.out.println("DUMMY_DEPOT" + i + " X: " + dummyDepot.getAxisX());
			System.out.println("DUMMY_DEPOT" + i + " Y: " + dummyDepot.getAxisY());
	
			System.out.println("CENTROIDE" + i + " X: " + centroids.get(i).getLocationDepot().getAxisX());
			System.out.println("CENTROIDE" + i + " Y: " + centroids.get(i).getLocationDepot().getAxisY());
			
			if((centroids.get(i).getLocationDepot().getAxisX() != dummyDepot.getAxisX()) || (centroids.get(i).getLocationDepot().getAxisY() != dummyDepot.getAxisY())) 
			{
				change = true;
				
				centroids.get(i).setIDDepot(-1);
				
				Location location = new Location();
				location.setAxisX(dummyDepot.getAxisX());
				location.setAxisY(dummyDepot.getAxisY());
				centroids.get(i).setLocationDepot(location);	
				
				System.out.println("change: " + change);
				System.out.println("CENTROIDE" + i + " X: " + centroids.get(i).getLocationDepot().getAxisX());
				System.out.println("CENTROIDE" + i + " Y: " + centroids.get(i).getLocationDepot().getAxisY());
			}
			else
			{
				System.out.println("CENTROIDE" + i + " X: " + centroids.get(i).getLocationDepot().getAxisX());
				System.out.println("CENTROIDE" + i + " Y: " + centroids.get(i).getLocationDepot().getAxisY());
			}
		}	

		if(change)
			updateCentroids(clusters, centroids, distanceType);

		System.out.println("CAMBIO LOS CENTROIDES: " + change);
		
		return change;
	}
	
	private void updateCentroids(ArrayList<Cluster> clusters, ArrayList<Depot> centroids, DistanceType distanceType) {		
		NumericMatrix costMatrix = new NumericMatrix();
		try {
			costMatrix = Problem.getProblem().calculateCostMatrix(centroids, Problem.getProblem().getDepots(), distanceType);
		} catch (IllegalArgumentException | SecurityException
				| ClassNotFoundException | InstantiationException
				| IllegalAccessException | InvocationTargetException
				| NoSuchMethodException e) {
			e.printStackTrace();
		}
		
		ArrayList<Depot> tempCentroids = new ArrayList<Depot>(centroids);
		
		int totalCentroids = centroids.size();
		RowCol rcBestAll = new RowCol();
		int posCentroid = -1;
		int posDepot = -1;

		System.out.println("-------------------------------------" );
		for(int i = 0; i < centroids.size(); i++)
		{
			System.out.println("CENTROIDE ID: " + centroids.get(i).getIDDepot());
			System.out.println("CENTROIDE X: " + centroids.get(i).getLocationDepot().getAxisX());
			System.out.println("CENTROIDE Y: " + centroids.get(i).getLocationDepot().getAxisY());		
		}
		
		for(int i = 0; i < costMatrix.getRowCount(); i++)
		{
			for(int j = 0; j < costMatrix.getColCount(); j++){
				System.out.println("Row: " + i + " Col: " + j + " VALUE: " + costMatrix.getItem(i, j));
			}
			System.out.println("---------------------------------------------");
		}		
		
		while(!costMatrix.fullMatrix(0, 0, (totalCentroids  - 1), (totalCentroids - 1), Double.POSITIVE_INFINITY))
		{
			rcBestAll = costMatrix.indexLowerValue();

			System.out.println("BestAllRow: " + rcBestAll.getRow());
			System.out.println("BestAllCol: " + rcBestAll.getCol());
			System.out.println("COSTO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));
			
			posCentroid = rcBestAll.getRow();
			posDepot = rcBestAll.getCol();
			
			System.out.println("POSICIÓN DEL CENTROIDE: " + posCentroid);
			System.out.println("POSICIÓN DEL DEPOSITO: " + posDepot);

			if(posCentroid != posDepot)
			{
				Depot depot = new Depot();

				depot.setIDDepot(tempCentroids.get(posCentroid).getIDDepot());
				System.out.println("ID CENTROIDE: " + tempCentroids.get(posCentroid).getIDDepot());
				
				double axisX = 0.0; 
				double axisY = 0.0; 
				axisX = tempCentroids.get(posCentroid).getLocationDepot().getAxisX();
				axisY = tempCentroids.get(posCentroid).getLocationDepot().getAxisY();
				
				Location location = new Location();
				location.setAxisX(axisX);
				location.setAxisY(axisY);
				depot.setLocationDepot(location);
				
				ArrayList<Fleet> fleet = new ArrayList<Fleet>();
				fleet.addAll(tempCentroids.get(posCentroid).getFleetDepot());
				depot.setFleetDepot(fleet);
			
				centroids.set(posDepot, depot);
			}

			costMatrix.fillValue(0, posDepot, (totalCentroids - 1), posDepot, Double.POSITIVE_INFINITY);
			costMatrix.fillValue(posCentroid, 0, posCentroid, (totalCentroids - 1), Double.POSITIVE_INFINITY);
			
			for(int i = 0; i < costMatrix.getRowCount(); i++)
			{
				for(int j = 0; j < costMatrix.getColCount(); j++){
					System.out.println("Row: " + i + " Col: " + j + " VALUE: " + costMatrix.getItem(i, j));
				}
				System.out.println("---------------------------------------------");
			}		
		}
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/clustering/partitional/by_centroids.py
import numpy as np
from typing import List
from .partitional import Partitional
from ....service.distance_type import DistanceType
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.input.fleet import Fleet
from ....problem.input.location import Location

class ByCentroids(Partitional):
    
    def __init__(self):
        super().__init__()
    
    # Método que verifica y actualiza los centroides de los clústeres.
    def verify_centroids(self) -> bool:
        change: bool = False
        dummy_depot: Location = None
        
        print(f"change: {change}")
        
        for i, cluster in enumerate(self.list_clusters):
            if cluster.get_items_of_cluster():  # Si el clúster no está vacío
                dummy_depot = self.recalculate_centroid(cluster)
            else:
                dummy_depot = self.list_centroids[i].get_location_depot()
            
            print("------------------------------------------------------------------")
            print(f"DUMMY_DEPOT {i} X: {dummy_depot.get_axis_x()}")
            print(f"DUMMY_DEPOT {i} Y: {dummy_depot.get_axis_y()}")

            print(f"CENTROIDE {i} X: {self.list_centroids[i].get_location_depot().get_axis_x()}")
            print(f"CENTROIDE {i} Y: {self.list_centroids[i].get_location_depot().get_axis_y()}")
            
            if (self.list_centroids[i].get_location_depot().get_axis_x() != dummy_depot.get_axis_x() or self.list_centroids[i].get_location_depot().get_axis_y() != dummy_depot.get_axis_y()):
                change = True
                self.list_centroids[i].set_id_depot(-1)
                
                location = Location()
                location.set_axis_x(dummy_depot.get_axis_x())
                location.set_axis_y(dummy_depot.get_axis_y())
                self.list_centroids[i].set_location_depot(location)

                print(f"change: {change}")
                print(f"CENTROIDE {i} X: {self.list_centroids[i].get_location_depot().get_axis_x()}")
                print(f"CENTROIDE {i} Y: {self.list_centroids[i].get_location_depot().get_axis_y()}")
            else:
                print(f"CENTROIDE {i} X: {self.list_centroids[i].get_location_depot().get_axis_x()}")
                print(f"CENTROIDE {i} Y: {self.list_centroids[i].get_location_depot().get_axis_y()}")

        if change:
            self.update_centroids(self.list_centroids, self.distance_type)
            
        print(f"CAMBIO LOS CENTROIDES: {change}")
        
        return change
    
    # Método que actualiza los centroides de los clústeres.
    def update_centroids(
        self, 
        centroids: List[Depot], 
        distance_type: DistanceType
    ):
        cost_matrix: np.ndarray = None
        
        if distance_type == DistanceType.REAL:
            try:
                cost_matrix = Problem.get_problem().calculate_cost_matrix_real(
                    centroids, Problem.get_problem().get_depots()
                )
            except (ValueError, AttributeError, TypeError) as e:
                print("Error calculating cost matrix real:", e)
        else:
            try:
                cost_matrix = Problem.get_problem().calculate_cost_matrix(
                    centroids, Problem.get_problem().get_depots(), distance_type
                )
            except (ValueError, AttributeError, TypeError) as e:
                print("Error calculating cost matrix:", e)
        
        temp_centroids = centroids.copy()
        total_centroids = len(centroids)
        pos_centroid = -1
        pos_depot = -1
        
        print("-------------------------------------")
        for i, centroid in enumerate(centroids):
            print(f"CENTROIDE ID: {centroid.get_id_depot()}")
            print(f"CENTROIDE X: {centroid.get_location_depot().get_axis_x()}")
            print(f"CENTROIDE Y: {centroid.get_location_depot().get_axis_y()}")

        for i in range(cost_matrix.shape[0]):
            for j in range(cost_matrix.shape[1]):
                print(f"Row: {i} Col: {j} VALUE: {cost_matrix[i, j]}")
            print("---------------------------------------------")
            
        while not np.all(cost_matrix[:total_centroids, :total_centroids] == float('inf')):
            min_index = np.argmin(cost_matrix[:total_centroids, :total_centroids])
            row = min_index // cost_matrix.shape[1]
            col = min_index % cost_matrix.shape[1]

            print(f"BestAllRow: {row}")
            print(f"BestAllCol: {col}")
            print(f"COSTO: {cost_matrix[row, col]}")

            pos_centroid = row
            pos_depot = col

            print(f"POSICIÓN DEL CENTROIDE: {pos_centroid}")
            print(f"POSICIÓN DEL DEPÓSITO: {pos_depot}")

            if pos_centroid != pos_depot:
                depot = Depot()

                depot.set_id_depot(temp_centroids[pos_centroid].get_id_depot())
                print(f"ID CENTROIDE: {temp_centroids[pos_centroid].get_id_depot()}")

                axis_x = temp_centroids[pos_centroid].get_location_depot().get_axis_x()
                axis_y = temp_centroids[pos_centroid].get_location_depot().get_axis_y()

                location = Location()
                location.set_axis_x(axis_x)
                location.set_axis_y(axis_y)
                depot.set_location_depot(location)

                fleet: List[Fleet] = temp_centroids[pos_centroid].get_fleet_depot()
                depot.set_fleet_depot(fleet)

                centroids[pos_depot] = depot

            cost_matrix[:, pos_depot] = float('inf')
            cost_matrix[pos_centroid, :] = float('inf')

            for i in range(cost_matrix.shape[0]):
                for j in range(cost_matrix.shape[1]):
                    print(f"Row: {i} Col: {j} VALUE: {cost_matrix[i, j]}")
                print("---------------------------------------------")